GCTC-NTGC/gc-digital-talent
# Improve session-timeout logout experience

Resolves #4295.

~NOT YET FULLY WORKING.~
It's working now! 👍 

Steps to test:
1. Reduce the expiry timeout in mock-ouath config json and then restart docker container. I recommend setting it to 15 seconds.

a) Test restricted page with browser refresh

1. Log in and go to a page that requires being authorized as a particular user, like Profile page.
2. In the browser inspector, go to the Application tab, open localStorage, and delete the refresh token. (This will ensure that token refresh fails when the access token expires.)
3. Wait long enough for the access token to expire.
4. Refresh the browser page. You should be redirected to the mock-auth login page, then sent back to the profile page when login is successful.

b) Test restricted page with internal routing. 
1. Log in and go to a page that requires being authorized as a particular user, like Profile page.
2. In the browser inspector, go to the Application tab, open localStorage, and delete the refresh token. (This will ensure that token refresh fails when the access token expires.)
3. Wait long enough for the access token to expire.
4. Click a link to another page which requires being logged in (like the My Applications page). You should be redirected to the mock-auth login page, then sent back to the My Applications page which you were trying to navigate to.

c) Test public page with browser refresh
1. Log in and go to a page that does NOT require being authorized as a particular user, like home page.
2. In the browser inspector, go to the Application tab, open localStorage, and delete the refresh token. (This will ensure that token refresh fails when the access token expires.)
3. Wait long enough for the access token to expire.
4. Refresh the browser page. You should be logged out (indicate by changes to the nav bar) and then land back on the page you were on.

d) Test public page with internal routing
1. Log in and go to a page that does NOT require being authorized as a particular user, like home page.
2. In the browser inspector, go to the Application tab, open localStorage, and delete the refresh token. (This will ensure that token refresh fails when the access token expires.)
3. Wait long enough for the access token to expire.
4. Click a link to another page which does not require being logged in (like the Search page). You should be logged out (indicate by changes to the nav bar) and then land back on the page you were trying to reach.

- [x] Test a) passes
- [x] Test b) passes
- [x] Test c) passes
- [x] Test d) passes
- [x] Test a) passes when starting from an admin page
- [x] test b) passes when starting from an admin page

TODO:
~Right now, expected behaviour happens when you refresh the browser. However, if you trigger an api call without refreshing the browser after the access token has expired (ie tests b) and d)) then you get 401 unauthorized responses from the API, and console errors in the browser, but you are never actually logged out. I'm not sure why. Furthermore, if you reload the browser AFTER getting these errors, you still get the errors instead of proper behaviour.~
Resolved this issue! It turned out that an error in refreshTokenSet was causing code to crash before expected code could execute. This is fixed in acfcfab99e01e738b459376000f7a3f668161654.

QUESTION:
Oh, what should happen if you are on a restricted page (either the profile, or an admin), your session expires, you get redirected to logout, and then you log in as a different user (who's not an admin)?

A: The default behaviour on admin makes sense - you ge an error notice saying you're not authorized to view this page (because you don't have the correct role). I found a bit of an odd case on the profile. After logging as someone new, you're redirected to the previous user's profile, it seems to work just fine! Upon inspection, you'll notice the date is all from the new user, even though the url reflects the previous user's id, because we just use `getMe` graphql queries. I'll record this as a bug in another issue.
